### PR TITLE
[CARBONDATA-3581] Support page level bloom filter

### PR DESCRIPTION
**Pull request by kevinjmh**
**Target Scenario**
Data of column page: `1 2 3 4 99` => min is 1 and max is 99
Point Query: `column_value = 11` => minmax will tell you to scan this page, but actually no need.

This pr is introducing Page Level Bloom Filter when filter is applied on *column with discrete values*, which helps to skip scanning data for better performance. Page pruning runs in executor side.

**Support Datatype**
Currently only allow to build page bloom on column with datatype `int`, `bigint` or `string`, which resonable to have discrete values.

**Usage**
Specify columns in table property `page_bloom_include`. 
Page bloom filter will build for those columns when loading. 
Query will use the filter if filter columns has page bloom data and the filter is EQUAL or IN.

## Test
**Load time & Storage size for different table property**
`≈88 million rows`
`Property 1: 'page_bloom_include'='str1,bigint1'  `
`Property 2: 'sort_scope'='local_sort','sort_column'='str1,bigint1'  ` 

table property| loadTime(seconds) |  dataSize(GB)
-|-|-
nosort|2157.345|24.10
nosort + page bloom|2178.867|24.35
sort|4435.986 / 4281.805|16.65
sort + page bloom|4388.571 / 4220.84|16.93

Query: avg time cost of 10 times each case

type | 1 filter 1 project(seconds) |  1 filter 100 project(seconds)
-|-|-
nosort|2.1978 | 3.4139
nosort_bloom|1.1871 | 2.1172
sort|0.3114 | 0.5209
sort_bloom|0.3151 | 0.5164

cpu usage with same order as the form above
image

**Load time & Storage size for different number of page bloom column**

`no sort; ≈2.7 million rows;`

number of page bloom column | loadTime(seconds) |  dataSize(MB)
-|-|-
 1 | 58.998 | 777.92
 2 | 58.424 | 783.37
 3 | 58.197 | 785.14
 4 | 58.646 | 786.74
 5 | 58.621 | 788.67
 6 | 58.235 | 789.52
 7 | 58.954 | 791.29
 8 | 59.904 | 796.73
 9 | 58.785 | 802.17
 10 | 59.518 | 802.21
 11 | 59.051 | 802.23
 12 | 58.689 | 802.45
 13 | 59.06 | 802.66
 14 | 58.625 | 802.69
 15 | 59.973 | 808.13
 16 | 59.294 | 813.57
 17 | 59.08 | 813.58
 18 | 60.669 | 819.03
 19 | 59.484 | 824.46
 20 | 60.051 | 829.9
 21 | 59.728 | 830.98
 22 | 58.75 | 831.02
 23 | 59.612 | 831.03
 24 | 60.257 | 831.03
 25 | 59.42 | 831.78
 26 | 60.261 | 832.22
 27 | 59.843 | 832.23
 28 | 59.515 | 832.24
 29 | 61.135 | 833.27
 30 | 60.129 | 833.32
 31 | 60.393 | 833.5
 32 | 60.345 | 835.05
 33 | 59.882 | 840.49
 34 | 60.342 | 845.94
 35 | 60.656 | 851.38
 36 | 60.26 | 856.82
 37 | 60.764 | 862.27
 38 | 60.684 | 867.71
 39 | 60.539 | 869.84
 40 | 60.818 | 872.13
 41 | 60.743 | 872.14
 42 | 60.408 | 872.15
 43 | 61.21 | 872.15
 44 | 60.849 | 872.16
 45 | 63.1 | 873.6
 46 | 60.52 | 879.05
 47 | 60.701 | 879.05
 48 | 60.842 | 879.22
 49 | 62.261 | 879.52
 50 | 62.293 | 879.53

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/821/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/818/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/814/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/829/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/834/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/837/

**Comment by kevinjmh** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/854/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/860/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/862/

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 1]
```diff
@@ -501,6 +506,32 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
...
+    if (pageBloomFields.length != pageBloomFields.distinct.length) {
+      throw new MalformedCarbonCommandException(
+        CarbonCommonConstants.PAGE_BLOOM_INCLUDE.toUpperCase() + " have duplicate columns")
+    }
+    pageBloomFields.foreach { column =>
+      val bloomFiled = fields.find(_.column.equalsIgnoreCase(column))
+      if (bloomFiled.isDefined) {
+        // check if data type is supported
+        val datatypeStr = bloomFiled.get.dataType.getOrElse("");
+        val supportedType = Array("string", "int", "bigint")
```
Comment (ajantha-bhat) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: why not `short`  and other columns ?

File: core/src/main/java/org/apache/carbondata/core/bloom/BloomFilterUtil.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+   */
+  public static int[] getBloomParameters(int n, double p) {
+    double numOfBit = -n * Math.log(p) / (Math.pow(Math.log(2), 2));
+    double numOfHash = numOfBit / n * Math.log(2);
+    return new int[]{(int) Math.ceil(numOfBit), (int) Math.ceil(numOfHash)};
+  }
+
+  public static int[] getPageBloomParameters() {
+    return getBloomParameters(
+            CarbonV3DataFormatConstants.NUMBER_OF_ROWS_PER_BLOCKLET_COLUMN_PAGE_DEFAULT,
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/bloom/BloomFilterUtil.java]: we have `table_page_size_inmb` table properties, with that column page will not always have 3200 rows, it can have less than that also. 
Number of rows in a page is stored in thrift, `BlockletInfo3.row_count_in_page`. Need to create bloom vector based on this ?

**Comment by ajantha-bhat** [comment]
@kevinjmh : Any benchmarking is done for this?
How much store size increased and how much query performance improved ?

File: core/src/main/java/org/apache/carbondata/core/bloom/BloomFilterUtil.java (hunk 2 above)
Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/bloom/BloomFilterUtil.java]: You're right. I did try dynamic bloom parameter by dynamic size before. The benefit of setting the maximum pagesize(32000) are:
1. avoid problem of multiple parameters choosing as bloom datamap, only FPP needed to be set;
2. same parameter of bloom has same hash function, so that we only need to hash one time then we can check all pages in blocklet;

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala (hunk 1 above)
Comment (kevinjmh) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: I think these datatypes are more resonable to have discrete values and could be used as `col1 = xxx` or `col1 in (xxx, xxx)`

As for `short`, in a page of 32000 rows, that could easily we hit the page and tell executor to scan, and we waste time in bloom checking.

**Comment by kevinjmh** [comment]
> @kevinjmh : Any benchmarking is done for this?
> How much store size increased and how much query performance improved ?

Store size increased 1% when I set a string column and a bigint column to build page bloom.
Query performance is about 33.3% better in condition of no sort for point query, since many pages are skipped.

**Comment by ajantha-bhat** [comment]
>Store size increased 1% when I set a string column and a bigint column to build page bloom.
Query performance is about 33.3% better in condition of no sort for point query, since many pages are skipped.

What was the size of the data ? and  did you benchmark with local sort also ?

**Comment by kevinjmh** [comment]
> > Store size increased 1% when I set a string column and a bigint column to build page bloom.
> > Query performance is about 33.3% better in condition of no sort for point query, since many pages are skipped.
> 
> What was the size of the data ? and did you benchmark with local sort also ?

About 88milion rows. I did compare with local sort. Local sort only ask for several tasks but no sort ask a lot, so local sort in quite fast.  In a case, local sort cost 0.9s, no sort cost 3.0s and no sort with page bloom cost 2.0s.

**Comment by xuchuanyin** [comment]
Not quite sure about the performance cost & gain in query and loading scenarios with/without this patch.

**Comment by ajantha-bhat** [comment]
@kevinjmh : Also this PR needs a document update for new table property, please handle

File: processing/src/main/java/org/apache/carbondata/processing/store/TablePage.java [hunk 3]
```diff
@@ -104,6 +105,16 @@
     noDictDimensionPages = new ColumnPage[model.getNoDictionaryCount()];
     int tmpNumDictDimIdx = 0;
     int tmpNumNoDictDimIdx = 0;
+
+    // find columns config to build page bloom
+    List<String> pageBloomColumns = new ArrayList<>();
```
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/processing/store/TablePage.java]: No need to keep this new list to store bloom information. Instead in `DimensionSpec `and `MeasureSpec` add a flag for isBloomPage

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 4]
```diff
@@ -3158,6 +3158,28 @@ private static int unsetLocalDictForComplexColumns(List<ColumnSchema> allColumns
     return columnLocalDictGenMap;
   }
 
+
+  public static void setPageBloomColumnsToWrapperSchema(List<ColumnSchema> columns,
+        Map<String, String> mainTableProperties) {
+    if (!mainTableProperties.containsKey(CarbonCommonConstants.PAGE_BLOOM_INCLUDE)) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: here checking contains and down calling .get method, instead just call get method and validate for null

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 5]
```diff
@@ -3158,6 +3158,28 @@ private static int unsetLocalDictForComplexColumns(List<ColumnSchema> allColumns
     return columnLocalDictGenMap;
   }
 
+
+  public static void setPageBloomColumnsToWrapperSchema(List<ColumnSchema> columns,
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Add method level comments

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java [hunk 6]
```diff
@@ -249,7 +274,7 @@ public BitSet prunePages(RawBlockletColumnChunks rawBlockletColumnChunks)
   public boolean applyFilter(RowIntf value, int dimOrdinalMax) {
     if (isDimensionPresentInCurrentBlock) {
       byte[][] filterValues = dimColumnExecuterInfo.getFilterKeys();
-      byte[] col = (byte[])value.getVal(dimColumnEvaluatorInfo.getDimension().getOrdinal());
+      byte[] col = (byte[]) value.getVal(dimColumnEvaluatorInfo.getDimension().getOrdinal());
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: revert this change

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/column/ColumnSchema.java [hunk 7]
```diff
@@ -131,6 +131,11 @@
    */
   private String timeSeriesFunction = "";
 
+  /**
+   * set whether the column enables page bloom or not.
+   */
+  private boolean isPageBloomColumn = false;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/column/ColumnSchema.java]: need to handle in `readFields `and `writeFields `method for seralization and deseralization

**Comment by kumarvishal09** [comment]
@kevinjmh Can u please check with more number of columns as bloom page column(at least 20 to 50) and share the stats how much store size is increased

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala (hunk 1 above)
Comment (kumarvishal09) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: @kevinjmh Your comment is not clear, Can u please explain why you cannot support for short data type and other data type

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java [hunk 8]
```diff
@@ -54,6 +55,15 @@ public AbstractRawColumnChunk(int columnIndex, ByteBuffer rawData, long offSet,
     this.length = length;
   }
 
+  public ColumnPagesBloomFilter getPageBloomFilter() {
+    // page bloom is disabled for current column in this blocklet
+    if (!dataChunkV3.isSetPage_bloom_chunk()) {
+      return null;
+    } else {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java]: no need of else statement directly add return new ColumnPagesBloomFilter(dataChunkV3.page_bloom_chunk);

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java [hunk 9]
```diff
@@ -217,6 +220,14 @@ public BitSet prunePages(RawBlockletColumnChunks rawBlockletColumnChunks)
           bitSet.set(i);
         }
       }
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: when `carbon.push.rowfilters.for.vector` = `true`,  This method won't be called. 
Need to handle same in ```
applyFilter(RawBlockletColumnChunks rawBlockletColumnChunks,
      boolean useBitsetPipeLine)
```

File: format/src/main/thrift/carbondata.thrift [hunk 10]
```diff
@@ -151,8 +151,16 @@ struct DataChunk3{
...
     4: optional LocalDictionaryChunk local_dictionary; // to store blocklet local dictionary values
-   
+    5: optional PageBloomChunk page_bloom_chunk; // data of hash function and bitmaps of each page
  }
+
+ struct PageBloomChunk{
+    1: required i32 vector_size;
+    2: required i16 num_hash;
+    3: required i16 hash_type;
+    4: required list<binary> pagebloom_list;
```
Comment (kumarvishal09) [inline format/src/main/thrift/carbondata.thrift]: can u please write comment, And here one element in list represents one page bloom data ??

**Comment by ajantha-bhat** [comment]
@kevinjmh : Instead of one has algorithm, if we support multiple algorithm at once ? (need to store bitmap for each vector) , we can reduce the chances of false positive to much smaller. Any thoughts on this and tried benchmarking for this ?

File: core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+  /**
+   * prune pages based on minmax result bitset
+   */
+  public BitSet prunePages(byte[][] filterValues, BitSet minMaxBitSet) throws IOException {
+    // decode pages which is set to TRUE in minMaxBitSet
+    RoaringBitmap[] candidatePageBitmaps = new RoaringBitmap[minMaxBitSet.length()];
+    int length = minMaxBitSet.cardinality();
+    int nextSetBit = minMaxBitSet.nextSetBit(0);
+    for (int i = 0; i < length; ++i) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java]: Change this loop like below  
` for (int i = minMaxBitSet.nextSetBit(0); i >= 0; i = minMaxBitSet.nextSetBit(i + 1)) `{

File: core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      byte[] bloomFilterInBytes = bloomByteBuffers.get(nextSetBit).array();
+      DataInputStream dis = new DataInputStream(new ByteArrayInputStream(bloomFilterInBytes));
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java]: Please add detail level comment for this logic

File: core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+      RoaringBitmap bitmap = new RoaringBitmap();
+      bitmap.deserialize(dis);
+      candidatePageBitmaps[nextSetBit] = bitmap;
+      nextSetBit = minMaxBitSet.nextSetBit(nextSetBit + 1);
+    }
+
+    // check if page is hit in bloom
+    BitSet pageBitSet = new BitSet(minMaxBitSet.length());
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java]: @kevinjmh Please add detail level comments for below logic

File: core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java (hunk 12 above)
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/bloom/ColumnPagesBloomFilter.java]: Please close data input stream, here its not opening any file but then also its better to close the stream

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java (hunk 8 above)
Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java]: if page bloom is disabled for current column, the page_bloom_chunk will be NULL, no need to new a useless ColumnPagesBloomFilter object, and we still need to decide whether the object is efftive for pruning

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 9 above)
Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: I know that, as written in the description of this PR. However, applyFilter and prunePages is quite similar, I am wondering if anyone can refactor this part first.

File: format/src/main/thrift/carbondata.thrift (hunk 10 above)
Comment (kevinjmh) [reply format/src/main/thrift/carbondata.thrift]: yes

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java (hunk 8 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java]:
```suggestion
    if (!dataChunkV3.isSetPage_bloom_chunk()) {
      return null;
    } 
    return new ColumnPagesBloomFilter(dataChunkV3.page_bloom_chunk);
```

He means like above

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java]:
```suggestion
 if (!dataChunkV3.isSetPage_bloom_chunk()) {
      return null;
    } 
    return new ColumnPagesBloomFilter(dataChunkV3.page_bloom_chunk);
```
he means like above

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 9 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: I saw that description now. 

**Whether you do refactor or not need to handle in the apply filter now.** 

If possible you can refactor also. If not just handle your scenario , later we can refactor

File: core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java (hunk 8 above)
Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/datastore/chunk/AbstractRawColumnChunk.java]: OK, thanks @ajantha-bhat

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 9 above)
Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: ok I will handle in `applyFilter`

**Comment by kevinjmh** [comment]
> @kevinjmh : Instead of one has algorithm, if we support multiple algorithm at once ? (need to store bitmap for each algorithm) , we can reduce the chances of false positive to much smaller. Any thoughts on this and tried benchmarking for this ?

sorry, I'm not quite get the `algorithm` you mean.
currently we use Murmur Hash with multiple hash values for items added into bloom, position of the hash values is set to true. Query value will be considered inside the page only when all position of its hash vallue is true in the bitmap.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 9 above)
Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: I re-think about this.Different to minmax, page bloom costs more(storage, decode) when query. Row level filter may not benefit from bloom.

If we use **one filter column** and get multiple columns, once bloom says that page does not need to scan, nothing need to do for all columns of this page in direct fill case. If bloom can skip more pages, the IO benefit for skipped pages is `# of pages * project columns`. 

As for row filter, for same query, the benefit is shrank to `# of pages * 1 (the filter column)`. 

For row level filter:
```
1.original: 

decode page -> for-loop checking each value of this column

2.with page bloom: 

read bloom chunk -> decode bloom bitmap -> check bloom

if check result is false -> skip
if check result is true ->  decode page -> for-loop checking each value of this column
```

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/990/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/987/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/979/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/991/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/988/

**Comment by ajantha-bhat** [comment]
> @ajantha-bhat sorry, I'm not quite get the algorithm you mean.
currently we use Murmur Hash with multiple hash values for items added into bloom, position of the hash values is set to true. Query value will be considered inside the page only when all position of its hash vallue is true in the bitmap.

@kevinjmh : currently we use only one hash function (murmur hash) , result of it also may be false positive in case of hash collision. To improve this scenario. People use more than one hash function for same value (example fnv hash). Then if both the algorithm says true, then only it is considered to be exists.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 9 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: yeah. make sense. Let me check again

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: But it will be useful If data doesn't exist in page. we can skip the page, avoiding checking every row of that page.
However If the data exist in page. It is an extra operation.
This problem exist in prunePages() also. we decode bloom chunk , even though data exist  in page.

Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: That sounds like some gambling. Bloom check v.s. Direct scan. Skip pages v.s. Extra Operation.
Need some more test.

**Comment by kevinjmh** [comment]
I get that. But this may cost more for loading, storage and the gambling in query we talked above.

**Comment by ravipesala** [comment]
@kevinjmh Please check mailing list for my comments.
 http://apache-carbondata-dev-mailing-list-archive.1130556.n5.nabble.com/DISCUSSION-Page-Level-Bloom-Filter-td85720.html

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1191/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1209/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1199/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1192/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1210/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1200/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1244/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1254/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1263/

**Comment by wangyi-fudan** [comment]
just mention wyhash, the fastest high quality conventional hash function:
https://github.com/wangyi-fudan/wyhash

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/19/
